core/src/main/java/org/elasticsearch/action/DocWriteResponse.java

```diff
@@ -194,7 +200,17 @@ public String getLocation(@Nullable String routing) {
         if (routing != null) {
             location.append(routingStart).append(routing);
         }
-        return location.toString();
+
+        String locationString = location.toString();
+        try {
+            URI uri = new URI(locationString);
+            locationString = uri.toASCIIString();
+        } catch (URISyntaxException ex) {
+            // could not encode URI properly, using toString instead.
+            logger.debug("Location string is not a valid URI.");
```

Comment:
Could you add the exception to the log statement? If we really don't expect this to happen I think it should be a `warn` level. If someone is able to reproduce it then they can silence it with the update settings API before we work around it.

Now that I think about it, it'd be nicer to throw the `URISyntaxException` out of the method and then catch it when we go to write the location, warning there instead of here. That way we don't add the `Location` header with an empty string.